cmd/kaniko-acr/main.go

```diff
@@ -0,0 +1,388 @@
+			EnableCache:      c.Bool("enable-cache"),
+			CacheRepo:        fmt.Sprintf("%s/%s", c.String("registry"), c.String("cache-repo")),
+			CacheTTL:         c.Int("cache-ttl"),
+			DigestFile:       defaultDigestFile,
+			NoPush:           noPush,
+			Verbosity:        c.String("verbosity"),
+			Platform:         c.String("platform"),
+			SkipUnusedStages: c.Bool("skip-unused-stages"),
+		},
+		Artifact: kaniko.Artifact{
+			Tags:         c.StringSlice("tags"),
+			Repo:         c.String("repo"),
+			Registry:     c.String("registry"),
+			ArtifactFile: c.String("artifact-file"),
+			RegistryType: artifact.Docker,
+		},
+	}
+	return plugin.Exec()
+}
+
+func createDockerConfig(tenantId, clientId, certPath,
+	clientSecret, registry string, noPush bool) (*docker.Config, error) {
+	dockerConfig := docker.NewConfig()
+
+	if !noPush || clientId != "" {
+		// only setup auth when pushing or credentials are defined
+		if registry == "" {
+			return nil, fmt.Errorf("registry must be specified")
+		}
+
+		token, err := getAcrToken(tenantId, clientId, clientSecret, certPath, registry)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to fetch acrToken")
+		}
+		fmt.Printf("acr token " + token)
+		err = createDockerCfgFile(username, token, registry)
+	}
+
+	return dockerConfig, nil
+}
+
+func getAcrToken(tenantId, clientId, clientSecret, certPath, registry string) (string, error) {
+	if tenantId == "" {
+		return "", fmt.Errorf("tenantId can't be empty foe AAD authentication")
+	}
+
+	if clientId == "" {
+		return "", fmt.Errorf("clientId can't be empty foe AAD authentication")
+	}
+
+	if clientSecret == "" && certPath == "" {
+		return "", fmt.Errorf("one of accessKey or secretKey should be defined")
+	}
+
+	// TODO check for presence of file as well.
+	os.Setenv(accessKeyEnv, clientId)
+	os.Setenv(secretKeyEnv, clientSecret)
+	os.Setenv(tenantKeyEnv, tenantId)
+	env, err := azidentity.NewEnvironmentCredential(nil)
+	context.Background()
+	policy := policy.TokenRequestOptions{
+		Scopes: []string{"https://management.azure.com/.default"},
+	}
+
+	os.Unsetenv(accessKeyEnv)
+	os.Unsetenv(secretKeyEnv)
+	os.Unsetenv(tenantKeyEnv)
+
+	azToken, err := env.GetToken(context.Background(), policy)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch access token")
+	}
+
+	acrToken, err := fetchAcrToken(tenantId, azToken.Token, registry)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch acr token")
+	}
+	return acrToken, nil
+}
+
+func fetchAcrToken(tenantId, token, registry string) (string, error) {
+	formData := url.Values{
+		"grant_type":   {"access_token"},
+		"service":      {registry},
+		"tenant":       {tenantId},
+		"access_token": {token},
+	}
+	jsonResponse, err := http.PostForm(fmt.Sprintf("https://%s/oauth2/exchange", registry), formData)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch acr token")
+	}
+	var response map[string]interface{}
+	json.NewDecoder(jsonResponse.Body).Decode(&response)
```

Comment:
does it return err while decoding. If yes, handle it

Reply:
done